File: minimove/Assets/Gameflow.cs
using UnityEngine;
using System.Collections.Generic;
using System.Collections;


public class GameFlow {
	private static readonly System.Random rnd = new System.Random();
	List<MovePlayer> players = new List<MovePlayer>();
	MiniGame currentGame = null;
	MonoBehaviour behaviour = null;
	int remainingGames = 0;
	string currentMessage = "***";

	public GameFlow(GameObject gameObject, MonoBehaviour behaviour) {
		int count = UniMoveController.GetNumConnected ();
		Debug.Log ("Connected controllers: " + count);

		int player = 0;
		for (int i = 0; i < count; i++) {
			MovePlayer mp = new MovePlayer (gameObject, player);
			if (mp.Valid()) {
				player++;
				players.Add (mp);
			}
		}

		this.behaviour = behaviour;
		ResetGame ();
	}

	public void ResetGame() {
		this.remainingGames = GetTunables ().DefaultNumberOfGames;

		foreach (var player in players) {
			player.Score = 0;
		}

		SelectNewGame ();
	}

	public IEnumerator PlaySoundDelayed(string sound, float delaySec) {
		yield return new WaitForSeconds (delaySec);
		PlaySound (sound);
	}

	public void SelectNewGame() {
		// TODO: "New game starts sound"

		int maxScore = 0;
		foreach (var player in players) {
			player.LEDColor = Color.black;
			player.Rumble = 0f;
			maxScore = Mathf.Max (maxScore, player.Score);
		}
			
		Debug.Log("Selecting new game");
		if (remainingGames == 0) {
			Debug.Log ("Session ends");
			currentMessage = "Session ** Ends";
			int remainingWinners = 0;

			StartCoroutine (PlaySoundDelayed ("GameWin2Sound", GetTunables().GameWinAnimationWaitBeforeSec));

			foreach (var player in players) {
				OnFinished onFinished = delegate () {
					Debug.Log("onFinished called");
					remainingWinners -= 1;
					if (remainingWinners == 0) {
						ResetGame();
					}
				};
				Debug.Log ("Player score: " + player.Score + ", max score: " + maxScore);
				if (maxScore == player.Score) {
					remainingWinners += 1;
					StartCoroutine(player.GameWinAnimation (GetTunables (), onFinished));
				}
			}
		} else {
			List<MiniGame> games = new List<MiniGame> ();
			games.Add (new MoveSays (this));
			games.Add (new ShakeIt (this));
			games.Add (new Freeze (this));
			games.Add (new SafeCracker(this));

			MiniGame candidate = null;
			do {
				candidate = games [rnd.Next (games.Count)];
			} while (!candidate.CanSupportPlayers(players.Count));

			currentGame = candidate;
				
			currentGame.StartGame ();
			remainingGames--;
		}
	}

	void UpdateControllers() {
		/*
		Color[] colors = { Color.cyan, Color.red, Color.blue, Color.green, Color.magenta };
*/
		foreach (var player in players) {
			// Geht L2, L1, R1, ... (alle PSMoveButton-Werte) durch
			foreach (PSMoveButton button in System.Enum.GetValues(typeof(PSMoveButton))) {
				if (player.move.GetButtonDown (button)) {
					if (currentGame != null) {
						currentGame.ButtonPressed (player, button);
					}
				}
			}

			/*player.LEDColor = colors [player.PlayerNumber % colors.Length];

			if (player.move.GetButtonDown (PSMoveButton.Cross)) {
				Debug.Log ("Button pressed");
			}*/

			player.Update ();
		}
	}

	public IEnumerator DestroySoundLater(GameObject go, AudioSource src) {
		while (src.isPlaying) {
			yield return new WaitForSeconds (0.01f);
		}

		Object.Destroy (go);
	}
		
	public void PlaySound(string sound, float volume=1.0f, float pitch=1.0f) {
		GameObject soundObject = GameObject.Find (sound);
		AudioSource soundSource = soundObject.GetComponent<AudioSource> ();

		GameObject go = new GameObject ("Playing:" + sound);
		AudioSource src = go.AddComponent<AudioSource> ();
		src.volume = volume;
		src.pitch = pitch;
		src.clip = soundSource.clip;
		src.Play ();
		behaviour.StartCoroutine (DestroySoundLater (go, src));
	}

	public void Update () {
		UpdateControllers ();

		if (currentGame != null) {
			currentGame.Update ();
		}
	}

	public TunableVariables GetTunables() {
		return behaviour.GetComponent<TunableVariables> ();
	}

	public void endCurrentGameNoWinner() {
		endCurrentGame (new List<MovePlayer>() {});
	}

	public void endCurrentGame(MovePlayer winner) {
		endCurrentGame (new List<MovePlayer>() { winner });
	}

	public void endCurrentGame(List<MovePlayer> winners) {
		if (currentGame == null) {
			return;
		}

		// Disable access to "old" current game; new current game
		// will be set by the winner player 
		currentGame = null;


		Debug.Log ("Game ends, winner(s): " + winners);

		foreach (var player in players) {
			player.LEDColor = Color.black;
		}

		if (winners.Count == 0) {
			// TODO: Play "nobody wins" sound/animation and wait a bit before new game
			SelectNewGame ();
		} else {
			string[] winningSounds = new string[] { "WinPlayer1Sound", "WinPlayer2Sound" };

			PlaySound (winningSounds[winners[0].PlayerNumber % winningSounds.Length], 0.2f);

			OnFinished onFinished = delegate () {
				SelectNewGame();
			};

			foreach (var winner in winners) {
				//winner.LEDColor = Color.white;
				StartCoroutine (winner.WinAnimation (GetTunables (), onFinished));
				winner.Score++;
				onFinished = delegate() {
					// Do nothing
				};
			}
		}
	}

	public void StartCoroutine(IEnumerator coroutine) {
		behaviour.StartCoroutine (coroutine);
	}

	public List<MovePlayer> Players {
		get { return players; }
	}

	public string StatusMessage {
		get {
			string result = "";
			if (currentGame != null) {
				result += "current game: " + currentGame;
				result += "\n" + currentGame.StatusMessage ();
			} else {
				result += currentMessage;
			}

			result += "\n";
			foreach (var player in players) {
				result += "\n";
				result += "Player " + (player.PlayerNumber + 1) + ": " + player.Score + " points";
			}
			return result;
		}
	}
}

File: python/gameflow.py
import psmoveapi
import moveplayer
import coroutine
import tunablevariables

import random

from color import Color

from minigames import attractmode
from minigames import movesays
from minigames import shakeit
from minigames import safecracker


class Counter(object):
    def __init__(self):
        self.value = 0

    def increment(self):
        self.value += 1

    def decrement(self):
        self.value -= 1
        return self.value == 0


class GameFlow(object):
    def __init__(self, main_script):
        self.main_script = main_script

        self.players = [moveplayer.MovePlayer(index, controller)
                        for index, controller in enumerate(self.main_script.get_controllers())]

        self.current_game = None
        self.remaining_games = 0
        self.current_message = '***'

        self.reset_game()

    def reset_game(self):
        self.remaining_games = self.get_tunables().DefaultNumberOfGames

        for player in self.players:
            player.score = 0

        self.current_game = attractmode.AttractMode(self)
        self.current_game.start_game()

    def everyone_is_ready(self):
        self.current_game = None
        self.select_new_game()

    def play_sound_delayed(self, sound, delay_sec):
        yield coroutine.WaitForSeconds(delay_sec)
        self.play_sound(sound)

    def select_new_game(self):
        # TODO: "New game starts sound"
        tunables = self.get_tunables()

        max_score = 0
        for player in self.players:
            player.led_color = Color.BLACK
            player.rumble = 0.0
            max_score = max(max_score, player.score)

        if self.remaining_games == 0:
            self.current_message = 'Session ** Ends'
            remaining_winners = Counter()

            self.start_coroutine(self.play_sound_delayed('GameWin2Sound', tunables.GameWinAnimationWaitBeforeSec))

            for player in self.players:
                def on_finished():
                    if remaining_winners.decrement():
                        self.reset_game()
                if max_score == player.score:
                    remaining_winners.increment()
                    self.start_coroutine(player.game_win_animation(tunables, on_finished))
        else:
            games = []
            if tunables.EnableMoveSays:
                games.append(movesays.MoveSays(self))
            if tunables.EnableShakeIt:
                games.append(shakeit.ShakeIt(self))
            if tunables.EnableSafeCracker:
                games.append(safecracker.SafeCracker(self))

            self.current_game = random.choice(games)
            self.current_game.start_game()
            self.remaining_games -= 1

    def update_controllers(self):
        for player in self.players:
            # Geht L2, L1, R1, ... (alle PSMoveButton-Werte) durch
            for button in psmoveapi.Button.VALUES:
                if player.is_button_down(button):
                    if self.current_game is not None:
                        self.current_game.button_pressed(player, button)

            player.update()

    def play_sound(self, sound, volume=1.0, pitch=1.0):
        self.main_script.play_sound(sound, volume, pitch)

    def update(self):
        self.update_controllers()

        if self.current_game is not None:
            self.current_game.update()

    def get_tunables(self):
        return tunablevariables.TunableVariables

    def end_current_game_no_winner(self):
        self.end_current_game()

    def end_current_game(self, *winners):
        if self.current_game is None:
            return

        # Disable access to "old" current game; new current game will be set by the winner player
        self.current_game = None

        for player in self.players:
            player.led_color = Color.BLACK

        if len(winners) == 0:
            # TODO: Play "nobody wins" sound/animation and wait a bit before new game
            self.current_message = 'Nobody wins...'
            self.select_new_game()
        else:
            winning_sounds = ['WinPlayer1Sound', 'WinPlayer2Sound']

            if len(winners) == 1:
                self.current_message = 'Player %d wins the round' % (winners[0].player_number + 1)
            else:
                self.current_message = 'Winners: Player ' + ' and '.join(str(w.player_number+1) for w in winners)

            self.play_sound(winning_sounds[winners[0].player_number % len(winning_sounds)], 0.2)

            def on_finished():
                self.select_new_game()

            for winner in winners:
                self.start_coroutine(winner.win_animation(self.get_tunables(), on_finished))
                winner.score += 1
                def on_finished():
                    ...  # Do nothing

    def start_coroutine(self, crt):
        self.main_script.start_coroutine(crt)

    def status_message(self):
        result = []
        if self.current_game is not None:
            result.append(self.current_game.__class__.__name__ if self.current_game else '-')
            result.append(self.current_game.status_message())
        else:
            result.append('Woop!')
            result.append(self.current_message)

        result.append(str(player.score) for player in self.players)

        return result
